service/history/historyEngine.go

```diff
@@ -2796,7 +2802,66 @@ func (e *historyEngineImpl) ReapplyEvents(
 		ctx,
 		domainID,
 		currentExecution,
-		func(mutableState mutableState) (*updateWorkflowAction, error) {
+		func(context workflowExecutionContext, mutableState mutableState) (*updateWorkflowAction, error) {
+
+			if !mutableState.IsWorkflowExecutionRunning() {
+				// need to reset target workflow (which is also the current workflow)
+				// to accept events to be reapplied
+				baseRunID := mutableState.GetExecutionInfo().RunID
+				resetRunID := uuid.New()
+				baseRebuildLastEventID := mutableState.GetPreviousStartedEventID()
+
+				// TODO when https://github.com/uber/cadence/issues/2420 is finished, remove this block,
+				//  since cannot reapply event to a finished workflow which had no decisions started
+				if baseRebuildLastEventID == common.EmptyEventID {
+					e.logger.Warn("cannot reapply event to a finished workflow",
+						tag.WorkflowDomainID(domainID),
+						tag.WorkflowID(currentExecution.GetWorkflowId()),
+					)
+					e.metricsClient.IncCounter(metrics.HistoryReapplyEventsScope, metrics.EventReapplySkippedCount)
+					return &updateWorkflowAction{noop: true}, nil
+				}
+
+				baseVersionHistories := mutableState.GetVersionHistories()
+				baseCurrentVersionHistory, err := baseVersionHistories.GetCurrentVersionHistory()
+				if err != nil {
+					return nil, err
+				}
+				baseRebuildLastEventVersion, err := baseCurrentVersionHistory.GetEventVersion(baseRebuildLastEventID)
+				if err != nil {
+					return nil, err
+				}
+				baseCurrentBranchToken := baseCurrentVersionHistory.GetBranchToken()
+				baseNextEventID := mutableState.GetNextEventID()
+
+				if err = e.workflowResetter.resetWorkflow(
+					ctx,
+					domainID,
+					workflowID,
+					baseRunID,
+					baseCurrentBranchToken,
+					baseRebuildLastEventID,
+					baseRebuildLastEventVersion,
+					baseNextEventID,
+					resetRunID,
+					uuid.New(),
+					newNDCWorkflow(
+						ctx,
+						e.shard.GetDomainCache(),
+						e.shard.GetClusterMetadata(),
+						context,
+						mutableState,
+						noopReleaseFn,
+					),
+					eventsReapplicationResetWorkflowReason,
+					reapplyEvents,
+				); err != nil {
+					return nil, err
```

Comment:
If the workflow cannot be reset due to no decision task, will the reapply retry forever?

Reply:
line 2816 does the check and give up, until #2420 is finished